### PR TITLE
Verify token characters when added and when used

### PR DESCRIPTION
**Pull request by sigurdm**
Fixes: https://github.com/dart-lang/pub/issues/3725

### CODE REVIEW

File: doc/repository-spec-v2.md [hunk 1]
```diff
@@ -133,6 +133,10 @@ that the `<token>` is accidentally stored in shell history. For security reasons
 authentication can only be used when `<hosted-url>` uses HTTPS. For further
 details on token management see: `dart pub token --help`.
 
+The tokens are inserted verbatim in the header, therefore they have to adhere to
+ https://www.rfc-editor.org/rfc/rfc6750#section-2.1. This means they must match
+ the regex: `^[a-zA-Z0-9._~+/=-]*$`.
```
Comment (jonasfj) [inline doc/repository-spec-v2.md:138]: nit, but we should not allow empty tokens? Should we?
```suggestion
 the regex: `^[a-zA-Z0-9._~+/=-]+$`.
```

File: lib/src/authentication/credential.dart [hunk 2]
```diff
@@ -116,19 +117,26 @@ class Credential {
...
         );
       }
-      return Future.value('Bearer $value');
+      tokenValue = value;
+    } else {
+      tokenValue = token!;
+    }
+    if (!isValidBearerToken(tokenValue)) {
+      dataError('Credential token for $url is not a valid Bearer token. '
+          'It should match `^[a-zA-Z0-9._~+/=-]*\$`');
```
Comment (jonasfj) [inline lib/src/authentication/credential.dart:136]:
```suggestion
          'It should match `^[a-zA-Z0-9._~+/=-]+\$`');
```

File: lib/src/authentication/credential.dart [hunk 3]
```diff
@@ -144,6 +152,14 @@ class Credential {
   // Either [token] or [env] should be defined to be valid.
   bool isValid() => (token == null) ^ (env == null);
 
+  /// Whether [candidate] can be used as a bearer token.
+  ///
+  /// We limit tokens to be valid bearer tokens according to
+  /// https://www.rfc-editor.org/rfc/rfc6750#section-2.1
+  static bool isValidBearerToken(String candidate) {
+    return RegExp(r'^[a-zA-Z0-9._~+/=-]*$').hasMatch(candidate);
```
Comment (jonasfj) [inline lib/src/authentication/credential.dart:160]:
```suggestion
    return RegExp(r'^[a-zA-Z0-9._~+/=-]+$').hasMatch(candidate);
```

File: doc/repository-spec-v2.md (hunk 1 above)
Comment (jonasfj) [reply doc/repository-spec-v2.md:138]: Yeah, we don't allow empty tokens, see:
```
    if (token.isEmpty) {
      usageException('Token is not provided.');
    }
```

**Review by jonasfj — APPROVED** [review]
This is a potentially breaking change, let's make sure it's noted in CHANGELOG.

**Comment by sigurdm** [comment]
> This is a potentially breaking change,

Hmm, before you would get an http error about an invalid header... not sure how breaking it is to catch it before - but sure let's put it in the changelog.

**Pull request merged**